k8s/src/main/scala/io/buoyant/k8s/resources.scala

```diff
@@ -0,0 +1,308 @@
+
+  def namespace(ns: String) = new NsVersion[O](client, group, version, ns)
+  def path = s"/$group/$version"
+  def list[T <: O: Manifest, W <: Watch[T]: Manifest, L <: List[T]: Manifest](
+    backoffs: Stream[Duration] = Backoff.exponentialJittered(1.milliseconds, 5.seconds),
+    stats: StatsReceiver = DefaultStatsReceiver
+  )(implicit od: ObjectDescriptor[T, W]) = new ListResource[T, W, L](client, path, backoffs, stats)
+}
+
+/**
+ * A Version with a namespace applied. This is a Class because we want to provide a default `namespace` implementation
+ * in [[Version]], although in many cases implementors may wish to extend this to provide convenience methods wrapping
+ * `list`.
+ *
+ * @param client The client used to make requests against this API.
+ * @param version Full path for this version of the API.
+ * @param ns The namespace
+ * @tparam O The parent type for all [Object]s served by this version.
+ */
+private[k8s] class NsVersion[O <: Object](val client: Client, group: String, val version: String, ns: String) extends Resource {
+  def path = s"/$group/$version/namespaces/$ns"
+  def list[T <: O: Manifest, W <: Watch[T]: Manifest, L <: List[T]: Manifest](
+    backoffs: Stream[Duration] = Backoff.exponentialJittered(1.milliseconds, 5.seconds),
+    stats: StatsReceiver = DefaultStatsReceiver
+  )(implicit od: ObjectDescriptor[T, W]) = new NsListResource[T, W, L](client, path, backoffs, stats)
+}
+
+/**
+ * A third-party version contains ThirdPartyResource-typed objects. See
+ * https://github.com/kubernetes/kubernetes/blob/master/docs/design/extending-api.md for a description of the
+ * design model.
+ *
+ * Implementors are likely to want to extend this to provide convenience methods on top of the `list` method.
+ *
+ * @tparam O The parent type for all [Object]s served by this Version.
+ */
+trait ThirdPartyVersion[O <: Object] extends Version[O] {
+  /**
+   * domain name, i.e. "buoyant.io"
+   */
+  def owner: String
+
+  /**
+   * version within owner domain, i.e. "v1"
+   */
+  def ownerVersion: String
+
+  def version: String = ThirdPartyVersion.version(owner, ownerVersion)
+  def basePath: String = s"/${ThirdPartyVersion.group}/$version"
+  val group = ThirdPartyVersion.group
+  override def namespace(ns: String) = new NsThirdPartyVersion[O](client, owner, ownerVersion, ns)
+}
+
+object ThirdPartyVersion {
+  def version(owner: String, ownerVersion: String) = s"$owner/$ownerVersion"
+  val group = "apis"
+}
+
+/**
+ * A namespaced third-party version, i.e. "/apis/buoyant.io/v1/namespaces/ns/".
+ */
+class NsThirdPartyVersion[O <: Object](client: Client, owner: String, ownerVersion: String, ns: String)
+  extends NsVersion[O](client, ThirdPartyVersion.group, ThirdPartyVersion.version(owner, ownerVersion), ns)
+
+/**
+ * A trait that encapsulates the ability to Watch a k8s resource.
+ */
+private[k8s] trait Watchable[O <: Object, W <: Watch[O]] extends Resource {
+  implicit protected val watchManifest: Manifest[W]
```

Comment:
I made it a trait because it seems like something one might want to mix in to classes that have other superclasses, but you're right that in the current implementation it could be an abstract class.